### PR TITLE
Unit test for CRMP to discard entries with missing retained buffer

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem
Reliable message retransmission accesses retained message buffer without checking if it is valid. This leads to crash.

#### Change overview
Add a unit test for the problem scenario. The test is able to recreate the problem. Fix the problem by ignoring the retransmission entries with no message buffer.

#### Testing
`CheckFailedMessageRetainOnSend` is the new unit test that tests the code change.
Rest of CRMP unit tests (in `TestReliableMessageProtocol`) ensure that existing functionality is still intact.

### CODE REVIEW

File: src/transport/SecureSessionMgr.h [hunk 1]
```diff
@@ -94,13 +94,19 @@ class EncryptedPacketBufferHandle final : private System::PacketBufferHandle
     CHIP_ERROR InsertPacketHeader(const PacketHeader & aPacketHeader) { return aPacketHeader.EncodeBeforeData(*this); }
 #endif // CHIP_ENABLE_TEST_ENCRYPTED_BUFFER_API
 
+#ifdef CHIP_CONFIG_TEST
     void operator=(PacketBufferHandle && aBuffer) { PacketBufferHandle::operator=(std::move(aBuffer)); }
+#endif
```
Comment (bzbarsky-apple) [inline src/transport/SecureSessionMgr.h]:
```suggestion
#ifdef CHIP_CONFIG_TEST
  public:
#else
  private:
#endif // CHIP_CONFIG_TEST
    void operator=(PacketBufferHandle && aBuffer) { PacketBufferHandle::operator=(std::move(aBuffer)); }
```

is another option.  It ensures that we don't have different impls for tests and non-tests...

**Review by msandstedt — APPROVED** [review]

File: src/messaging/ReliableMessageMgr.cpp [hunk 2]
```diff
@@ -133,6 +133,13 @@ void ReliableMessageMgr::ExecuteActions()
         if (!rc || entry.nextRetransTimeTick != 0)
             continue;
 
+        if (entry.retainedBuf.IsNull())
+        {
+            // Discard the current entry if the retained buffer is null
```
Comment (andy31415) [inline src/messaging/ReliableMessageMgr.cpp:138]: The comment seems to re-read the code without telling me 'why' this is done. 
Could you rephrase to make the comment explain the code rather than re-reading it? Could also delete it, howver in this case I believe it may be valuable for the reader of the code to understand why entry may have an empty retained buffer and why clearing is approrpiate.

Could be `empty retaining buffer means CRMP has nothing to resend and entry is invalid` ... but is this correct? why would we have an entry with nextRetransTick set if the retained buffer is null?

File: src/transport/SecureSessionMgr.h [hunk 3]
```diff
@@ -92,13 +94,19 @@ class EncryptedPacketBufferHandle final : private System::PacketBufferHandle
     CHIP_ERROR InsertPacketHeader(const PacketHeader & aPacketHeader) { return aPacketHeader.EncodeBeforeData(*this); }
 #endif // CHIP_ENABLE_TEST_ENCRYPTED_BUFFER_API
 
+    // Making operator= available to tests by making it public.
```
Comment (andy31415) [inline src/transport/SecureSessionMgr.h:97]: Why is this needed? this looks very suspect especially with the `#ifdef`

**Review by andy31415 — CHANGES_REQUESTED** [review]
Could you find a solution without the `#ifdef` in the packetbufferhandle? That code looks very suspect.

File: src/transport/SecureSessionMgr.h (hunk 3 above)
Comment (andy31415) [reply src/transport/SecureSessionMgr.h:97]: I think making this public should be ok.

Comment (andy31415) [reply src/transport/SecureSessionMgr.h:97]: @yufengwangca - why where the constructors and assignment operators made private?

It seems we have this pattern of 'private constructor' and friend class which causes coupling isssues.

If we need to avoid casting things by mistake, how about providing a

```cpp
static EncryptedPacketBufferHandle MarkPacketEncrypted(PacketBufferHandle&& buffer);
```

which would make it clear that we have a packet that is encrypted and code was written on purpose to do this. I would completely drop the 'friend class'.

File: src/messaging/ReliableMessageMgr.cpp (hunk 2 above)
Comment (pan-apple) [reply src/messaging/ReliableMessageMgr.cpp:138]: I'll update the comment.

File: src/transport/SecureSessionMgr.h (hunk 3 above)
Comment (pan-apple) [reply src/transport/SecureSessionMgr.h:97]: I believe this was to ensure that Encrypted packet can only be created by SecureSessionMgr. I.e. to guarantee that the packet being retransmitted is correctly encrypted, and retransmit logic cannot accidentally send an unencrypted packet.

This, condition may be relaxed now, since for PASE/CASE session establishment messages, the retransmitted packet will not be encrypted.

I'll move it out of the ifdef condition.

Comment (andy31415) [reply src/transport/SecureSessionMgr.h:97]: I believe we should move away from coupling using friend classes.

Comment (bzbarsky-apple) [reply src/transport/SecureSessionMgr.h:97]: Just to give the history here: the idea was that becuase EncryptedPacketBufferHandle is transmitted without encryption it gives a way to transmit things without encryption in general if you can control the contents of the handle.  Which is not at all desirable. So all the ways to create an EncryptedPacketBufferHandle with random data that is not encrypted were marked private, and only the class that does encryption was allowed to do those things.

The reason this is coming up here is that in the test we _want_ to inject data of our choosing into an EncryptedPacketBufferHandle.  For one thing, that's the only way to test malformed data.  But ideally we would keep the API surface for doing this as small as possible, ideally test-only, and not expose it in general.

@andy31415 Any suggestions on how you would like to see this set of requirements addressed?

Comment (andy31415) [reply src/transport/SecureSessionMgr.h:97]: I believe having a 'MarkPacketBufferEncrypted' is acceptable as public and we should not have a friend class at all.

The class provides compile-type safety saying "you need an encrypted buffer here" and the developer has to explicitly mark buffers as encrypted.

Marking things private means that instead of `MarkEncrypted` a "hacky developer" would need to either change chip SDK sligtly (open source after all) or use some 'reinterpret cast' to change internal pointers. Either way, I do not want to fight developers and would rather have a clean API that is hard to misuse. I believe a 'mark is encrypted' API is clear enough and if misused it is so intentionally.

File: src/transport/SecureSessionMgr.h [hunk 4]
```diff
@@ -92,13 +97,18 @@ class EncryptedPacketBufferHandle final : private System::PacketBufferHandle
     CHIP_ERROR InsertPacketHeader(const PacketHeader & aPacketHeader) { return aPacketHeader.EncodeBeforeData(*this); }
 #endif // CHIP_ENABLE_TEST_ENCRYPTED_BUFFER_API
 
-private:
-    // Allow SecureSessionMgr to assign or construct us from a PacketBufferHandle
-    friend class SecureSessionMgr;
+    void MarkEncrypted(PacketBufferHandle && aBuffer)
```
Comment (andy31415) [inline src/transport/SecureSessionMgr.h:100]: This does both copy and mark.

Either `static EncryptedPacketBufferHandle MarkEncrypted(...)` and then you do:

``cpp
*bufferRetainSlot = EncryptedPacketBufferHandle::MarkEncrypted(msgBuf.Retain());
```

which by the way shows that we have an encrypted buffer that is NOT always marked as such (retain is a copy).

or have `void AssignFromEncrypted(...)` and then you do:

```cpp
bufferRetainSlot->AssignFromEncrypted(msgBuf.Retain());
```

I personally prefer the first version: one can create encrypted handles as long as an actual handle exists.

File: src/transport/SecureSessionMgr.h [hunk 5]
```diff
@@ -53,7 +53,7 @@ namespace chip {
  *  EncryptedPacketBufferHandle is a kind of PacketBufferHandle class and used to hold a packet buffer
  *  object whose payload has already been encrypted.
  */
-class EncryptedPacketBufferHandle final : private System::PacketBufferHandle
+class EncryptedPacketBufferHandle final : public System::PacketBufferHandle
```
Comment (andy31415) [inline src/transport/SecureSessionMgr.h:56]: why was the inheritence type changed?

File: src/transport/SecureSessionMgr.h [hunk 6]
```diff
@@ -92,13 +97,18 @@ class EncryptedPacketBufferHandle final : private System::PacketBufferHandle
...
+        mEncryptedPacketBuffer      = true;
+    }
+
+    bool IsEncrypted() const { return mEncryptedPacketBuffer; }
 
+private:
     EncryptedPacketBufferHandle(PacketBufferHandle && aBuffer) : PacketBufferHandle(std::move(aBuffer)) {}
 
-    void operator=(PacketBufferHandle && aBuffer) { PacketBufferHandle::operator=(std::move(aBuffer)); }
+    bool mEncryptedPacketBuffer = false;
```
Comment (andy31415) [inline src/transport/SecureSessionMgr.h:111]: The class name says "EncryptedPacketBufferHandle". Having a bool checking again if encrypted or not seems odd.

Can we update this in a different PR together with some renaming that makes the usage more logical?

**Review by andy31415 — APPROVED** [review]

**Comment by andy31415** [comment]
@bzbarsky-apple @jepenven-silabs @Damian-Nordic ?

File: src/transport/SecureSessionMgr.h (hunk 5 above)
Comment (bzbarsky-apple) [inline src/transport/SecureSessionMgr.h:56]: This was private so people couldn't accidentally start poking around at the buffer of an EncryptedPacketBufferHandle and modifying it.  Except SecureSessionMgr needs to do it, for the moment, hence it was a friend.

I would prefer that we keep this private, keep `SecureSessionMgr` as a friend, add `using System::PacketBufferHandle::IsNull;` so the null-test in reliable message manager can work....

Comment (bzbarsky-apple) [reply src/transport/SecureSessionMgr.h:56]: OK, talked to @andy31415 .  Let's land this as-is and I'll do a followup to put back in the safeguards.

**Review by bzbarsky-apple — APPROVED** [review]
